### PR TITLE
Range-ify std.file getTimes(), getTimesWin(), setTimes()

### PR DESCRIPTION
**Pull request by WalterBright**
http://dlang.org/phobos/std_file.html#.setTimes

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -767,13 +768,19 @@ void getTimes(in char[] name,
...
-            return stat(path.tempCString(), &buf);
+            return stat(namez, &buf);
         }
         stat_t statbuf = void;
 
-        cenforce(trustedStat(name, statbuf) == 0, name);
+        static if (isNarrowString!R && is(Unqual!(ElementEncodingType!R) == char))
+            alias names = name;
+        else
+            string names = null;
```
Comment (DmitryOlshansky) [inline std/file.d]: Ehm. Just convert range to string in cenfore? It's lazy parameter after all

File: std/file.d [hunk 2]
```diff
@@ -852,14 +859,17 @@ unittest
     Throws:
         $(D FileException) on error.
  +/
-version(StdDdoc) void getTimesWin(in char[] name,
+version(StdDdoc) void getTimesWin(R)(R name,
```
Comment (DmitryOlshansky) [inline std/file.d]: I totally expect horrible template code bloat with at least 3 functions per qualifier:
```
const(char)[]
immutable(char)[]
char[]
```

File: std/file.d [hunk 3]
```diff
@@ -982,27 +994,36 @@ void setTimes(in char[] name,
                          FILE_ATTRIBUTE_DIRECTORY |
                          FILE_FLAG_BACKUP_SEMANTICS,
                          HANDLE.init);
-        auto h = trustedCreateFileW(name, defaults);
+        auto h = trustedCreateFileW(namez, defaults);
 
-        cenforce(h != INVALID_HANDLE_VALUE, name);
+        static if (isNarrowString!R && is(Unqual!(ElementEncodingType!R) == char))
```
Comment (DmitryOlshansky) [inline std/file.d]: Again this miserable pattern, if anything it's a crude hack.

File: std/file.d (hunk 1 above)
Comment (WalterBright) [inline std/file.d]: That would mean that a string passed as an argument would have an extra copy of it allocated.

File: std/file.d (hunk 2 above)
Comment (WalterBright) [inline std/file.d]: Which is why for many of them I split out the implementation code into a separate function from the code that makes the 0 terminated argument.

File: std/file.d (hunk 3 above)
Comment (WalterBright) [inline std/file.d]: same answer :-)

File: std/file.d (hunk 1 above)
Comment (DmitryOlshansky) [inline std/file.d]: Only on exception and only if `name` it wasn't string to begin with. to! is really smart when it comes to (pun!) conversions of ee.g. string to string. The only problem I see is that it must be ForwardRange not input range to be able to use it twice.

Comment (DmitryOlshansky) [inline std/file.d]: Also cenforce prototype is:
```d
private T cenforce(T)(T condition, lazy const(char)[] name, string file = __FILE__, size_t line = __LINE__)
{
    if (!condition)
    {
      version (Windows)
      {
        throw new FileException(name, .GetLastError(), file, line);
      }
      else version (Posix)
      {
        throw new FileException(name, .errno, file, line);
      }
    }
    return condition;
}
```
What the heck is namez doing as 3rd parameter not to mention it being stack-allocated?!

File: std/file.d (hunk 3 above)
Comment (andralex) [inline std/file.d]: I do think long term we need to do something better about this. For now this is fine because it's internal only.

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**